elastic/elastic-agent
# Add action and state stores migrations

## What does this PR do?

Add migration for the action/state stores and the necessary tests.

Now it migrates:
 - from the action store
 - from the deprecated YAML state store

It also removes all action store unused code and creates the `internal/pkg/agent/storage/store/internal/migrations/` package. This package should contain all necessary and deprecated code to read the old stores data into current models. `internal/pkg/agent/storage/store/migrations.go` orchestrates the migrations.

## Why is it important?

The new State Store uses a different schema and is serialised in JSON instead of YAML

## Checklist




## How to test this PR locally

Testing action store -> new JSON state Store:
 - install a 7.17 agent
 - perform a policy update
 - ensure the action is persisted in the [action store](https://github.com/elastic/beats/blob/v7.17.18/x-pack/elastic-agent/pkg/agent/application/paths/files.go#L54). 
 - upgrade to this version
 - ensure the [action store](https://github.com/elastic/beats/blob/v7.17.18/x-pack/elastic-agent/pkg/agent/application/paths/files.go#L54) file is deleted and a new store.enc is present
 - [decrypt the encrypted](https://github.com/elastic/ingest-dev/pull/2937/files#diff-c71835d20a84e5014efc3a4cf8f31c3e2fb6e7111b25bdc718cc48a6671a8ed7R15-R41) store and check the same action is present.

Testing current state store -> new JSON state Store:
 - same as above, but start with the latest agent release
 - you might use a scheduled upgrade action to ensure there is an action in the action queue
 - the store will be encrypted by default, so:
 - [decrypt](https://github.com/elastic/ingest-dev/pull/2937/files#diff-c71835d20a84e5014efc3a4cf8f31c3e2fb6e7111b25bdc718cc48a6671a8ed7R15-R41) it
 - check the content
 - upgrade the agent
 - check the store content again. The information should be the same, but in JSON



## Related issues

- Relates #3912 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...
